qiskit/transpiler/passes/optimization/hoare_opt.py

```diff
@@ -260,10 +261,19 @@ def _target_successive_seq(self, qubit):
             except (IndexError, ValueError):
                 continue
 
-            if append:
-                seqs.append([node1, node2])
-
-        return seqs
+            seq = [node1, node2]
+            if append and self._is_identity(seq) and self._seq_as_one(seq):
+                i += 1
+                for node in seq:
+                    dag.remove_op_node(node)
+                    # if recursive call, gate will be removed further down
+                    if max_idx is None:
+                        for qbt in node.qargs:
+                            self.gatecache[qbt].remove(node)
+                    else:
+                        if self.gatecache[qubit].index(node) > max_idx:
+                            for qbt in node.qargs:
+                                self.gatecache[qbt].remove(node)
```

Comment:
A similar comment here - `_remove_successive_identity` isn't a recursive function.

Also, I know this was pre-existing, but this whole logic block is equivalent to
```python
if max_idx is None or self.gatecache[qubit].index(node) > max_idx:
    for qbt in node.qargs:
        self.gatecache[qbt].remove(node)
```
which avoids an extra scope, and removes some duplication.

(A comment far beyond the scope of the PR: this pre-existing code smells of poor scaling - `list.index` and `list.remove` are very tricky to use well, and this is highly likely to be at least quadratic in the circuit depth.)

Reply:
Indeed, I have merged the code as you suggested here.